test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,307 @@
+		assert.Exactly(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+	}
+	for _, vm := range vmList {
+		hash := sha1.New() // #nosec G401: not used for security purposes
+		io.WriteString(hash, vm.ifName)
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		t.Logf("Running verifyExternalEntityExistence")
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, true)
+		t.Logf("Running verifyExternalNodeRealization")
+		verifyExternalNodeRealization(vm)
+	}
+}
+
+func getVMInfo(t *testing.T, data *TestData, nodeName string) (info vmInfo) {
+	var err error
+	var vm vmInfo
+	vm.nodeName = nodeName
+	var cmd string
+	cmd = "ip -o -4 route show to default | awk '{print $5}'"
+	vm.osType = "Linux"
+	_, vm.ifName, _, err = data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v", cmd, nodeName, err)
+
+	vm.ifName = strings.TrimSpace(vm.ifName)
+	cmd = fmt.Sprintf("ifconfig %s | awk '/inet / {print $2}'| sed 's/addr://'", vm.ifName)
+	_, vm.ip, _, err = data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v", cmd, nodeName, err)
+
+	vm.ip = strings.TrimSpace(vm.ip)
+	return vm
+}
+
+func getWindowsVMInfo(t *testing.T, data *TestData, nodeName string) (vm vmInfo) {
+	var err error
+	vm.nodeName = nodeName
+	vm.osType = "Windows"
+	cmd := fmt.Sprintf("powershell 'Get-WmiObject -Class Win32_IP4RouteTable | Where { $_.destination -eq \"0.0.0.0\" -and $_.mask -eq \"0.0.0.0\"} | Sort-Object metric1 | select interfaceindex | ft -HideTableHeaders'")
+	_, ifIndex, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v", cmd, nodeName, err)
+
+	vm.ifIndex = strings.TrimSpace(ifIndex)
+	cmd = fmt.Sprintf("powershell 'Get-NetAdapter -IfIndex %s | select name | ft -HideTableHeaders'", vm.ifIndex)
+	_, ifName, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v", cmd, nodeName, err)
+
+	vm.ifName = strings.TrimSpace(ifName)
+	cmd = fmt.Sprintf("powershell 'Get-NetIPAddress -AddressFamily IPv4 -ifIndex %s| select IPAddress| ft -HideTableHeaders'", vm.ifIndex)
+	_, ifIP, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v", cmd, nodeName, err)
+
+	vm.ip = strings.TrimSpace(ifIP)
+	return vm
+
+}
+
+func startAntreaAgent(t *testing.T, data *TestData, vm vmInfo) {
+	t.Logf("Starting antrea-agent on VM: %s", vm.nodeName)
+	var cmd string
+	if vm.osType == "Windows" {
+		cmd = "nssm start antrea-agent"
+	} else {
+		cmd = "sudo systemctl start antrea-agent"
+	}
+	_, _, _, err := data.RunCommandOnNode(vm.nodeName, cmd)
```

Comment:
I noticed that most of the time you ignore stderr / stdout. Are these not useful for debugging in case of failure?

Reply:
The commands are the basic functions to start and stop antrea-agent. We do capture stdoout, whenever we need an output. for the failure cases, we do capture the err returned from the command, so I'm not sure if we need stderr output to be logged.

Reply:
> The commands are the basic functions to start and stop antrea-agent

Things can still go wrong.

If I am not mistaken, you use the remove provider and this will execute the command over SSH. `err` will only be `nil` if we fail to establish an SSH session. If the command run over SSH fails (i.e., returns a non-zero exit code), `err` will be `nil`. https://github.com/antrea-io/antrea/blob/main/test/e2e/providers/exec/ssh.go#L62-L64. That's why it's important to look at the first return value (exit code) and log stderr in case of a non-zero exit code.